plugin/v2/plugin_linux.go

```diff
@@ -136,5 +142,37 @@ func (p *Plugin) InitSpec(execRoot string) (*specs.Spec, error) {
 		p.modifyRuntimeSpec(&s)
 	}
 
+	// Rootless mode requires modifying the mount flags
+	// https://github.com/moby/moby/issues/47248#issuecomment-1927776700
+	// https://github.com/moby/moby/pull/47558
+	if userns.RunningInUserNS() {
+		for i := range s.Mounts {
+			m := &s.Mounts[i]
+			for _, o := range m.Options {
+				switch o {
+				case "bind", "rbind":
+					if _, err := os.Lstat(m.Source); err != nil {
+						if errors.Is(err, os.ErrNotExist) {
+							continue
+						}
+						return nil, err
+					}
+					// UnprivilegedMountFlags gets the set of mount flags that are set on the mount that contains the given
+					// path and are locked by CL_UNPRIVILEGED. This is necessary to ensure that
+					// bind-mounting "with options" will not fail with user namespaces, due to
+					// kernel restrictions that require user namespace mounts to preserve
+					// CL_UNPRIVILEGED locked flags.
+					//
+					// From https://github.com/moby/moby/blob/v23.0.1/daemon/oci_linux.go#L430-L460
+					unpriv, err := rootlessmountopts.UnprivilegedMountFlags(m.Source)
```

Comment:
Oh, wait! So we already have this in the `daemon/` but it's not exported, and ... BuildKit has a fork of that, and now we're pulling in the fork 🤔 

I see your TODO about considering moving this to `github.com/moby/sys`, but perhaps before that we could move the daemon `getUnprivilegedMountFlags` to our `internal/` package (at least until it's moved to a more permanent location).

Let me open a quick PR for that

Reply:
- opened https://github.com/moby/moby/pull/47567

Reply:
^ PR is merged (not sure if you're around at this time to update though)

Reply:
done